src/Groups/GAPGroups.jl

```diff
@@ -216,27 +216,126 @@ function rand_pseudo(G::GAPGroup; radius::Int = 10)
+# In this case, the parent of the resulting element is set according to
+# the following rules, depending on the types of the parents.
+#
+# - `PermGroup`, `PermGroup`:
+#   The operation is allowed whenever the parents have the same degree,
+#   and the parent is the symmetric group of that degree.
+#
+# - `PcGroup`, `PcGroup` and
+#   `FPGroup`, `FPGroup`:
+#   The operation is allowed whenever the parents belong to the same `GapObj`
+#   (in the sense of `==´), and the first of the two groups is taken.
+#
+# - `SubPcGroup`, `SubPcGroup` and
+#   `SubFPGroup`, `SubFPGroup`:
+#   The operation is allowed whenever the parents have the same `full_group`
+#   (in the sense of `==´), and the first of the two groups is taken.
+#
+# - `SubPcGroup`, `PcGroup` and
+#   `PcGroup`, `SubPcGroup` and
+#   `SubFPGroup`, `FPGroup` and
+#   `FPGroup`, `SubFPGroup`:
+#   The operation is allowed whenever the `full_group` of the `SubPcGroup`
+#   (`SubFPGroup`) is equal to the `PcGroup` (`FPGroup`)
+#   (in the sense of `==´), and the `PcGroup` (`FPGroup`) is taken.
+#
+# - `MatrixGroup`, `MatrixGroup`:
+#   The operation is allowed whenever the two groups have the same `degree`
+#   and `base_ring`, and the parent is the general linear group of that
+#   degree over that ring.
+#
+# - `AutomorphismGroup`, `AutomorphismGroup`:
+#   The operation is allowed whenever the two groups have the same `.G` field,
+#   and the parent is the full automorphism group of that group.
+#
+# - `DirectProductGroup`, `DirectProductGroup`:
+#   The operation is allowed whenever the two groups have the same `.L`
+#   field, and the parent is the direct product of these groups.
+#
+# For other types of groups, we throw an exception if the parent groups are
+# not equal and their `GapObj`s are not equal.
+#
+# Note that we cannot guarantee that the groups objects in question are
+# *identical*, only that their `GapObj`s are *equal*:
+# On the GAP side, the same group can be created several times,
+# and on the Oscar side, the same GAP group can be wrapped several times.
+#
+# Note also that we do not want to perform checks whether one group is perhaps
+# a subset of the other (and then one could take the bigger group as
+# the parent of the result).
+#
+function _common_parent_group(x::PermGroup, y::PermGroup)
+  x === y && return x
+  @req degree(x) == degree(y) "the groups have different degrees"
+  return symmetric_group(degree(x))
+end
+
+function _common_parent_group(x::PcGroup, y::PcGroup)
+  (x === y || GapObj(x) == GapObj(y)) && return x
+  throw(ArgumentError("the groups are not compatible"))
+end
+
+function _common_parent_group(x::FPGroup, y::FPGroup)
+  (x === y || GapObj(x) == GapObj(y)) && return x
+  throw(ArgumentError("the groups are not compatible"))
+end
+
+function _common_parent_group(x::SubPcGroup, y::SubPcGroup)
+  x === y && return x
+  @req x.full_group == y.full_group "the groups belong to different full groups"
+  return x.full_group
```

Comment:
So this is not type stable then, it sometimes gives a `SubPcGroup` and sometimes a `PcGroup`. Hrm, that's annoying, but I see no immediate way to resolve it, unless we were to make the "full group" deliberately a `SubPcGroup` that is an index 1 subgroup of the `PcGroup`...

Reply:
Taking that index 1 subgroups in `PcGroup`s occur already in many places because of these type conventions (the first member in `derived_series(G)` etc.), we can return always a `SubPcGroup`.
Currently the `SubPcGroup` stores the full `PcGroup`, thus fetching this `.full_group` is cheap, compared to wrapping a new `SubPcGroup`. Perhaps the `SubPcGroup` should store an index 1 subgroup instead?

Reply:
Hmm, if we want that a `SubPcGroup` stores a `SubPcGroup` as its `full_group` then we have the logical problem that the `SubPcGroup` constructor takes a `GapObj` and calls `SubPcGroup` for its full group, an infinite recursion ...